src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,382 @@
+
+        if (_mode == SymmetricCipherMode::GCM && _operation == Operation::Encryption) {
+            _authTag.resize(AuthenticatedAESCipher::AuthTagLength);
+            openssl::_EVP_CIPHER_CTX_ctrl(_ctx.get(), EVP_CTRL_GCM_GET_TAG,
+                                          AuthenticatedAESCipher::AuthTagLength,
+                                          _authTag.data());
+        }
+
+        processedChunk.resize(processingChunkSize);
+        _bufferStrategy->write(processedChunk);
+        _isFinished = true;
+
+        return _bufferStrategy->readAll();
+    }
+
+    std::vector<uint8_t> getIV() { return _iv; }
+
+    std::vector<uint8_t> getAuthTag() { return _authTag; }
+
+    void setAuthTag(const std::vector<uint8_t> &authTag) { _authTag = authTag; }
+
+private:
+    openssl::SSL_EVP_CIPHER_CTX_Ptr _ctx = nullptr;
+    std::vector<uint8_t> _iv;
+    std::vector<uint8_t> _authTag;
+    bool _isFinished = false;
+    Operation _operation;
+    SymmetricCipherMode _mode;
+    std::unique_ptr<CipherMemoryStrategyI> _bufferStrategy = nullptr;
+};
+
+
+AESCipher::AESCipher(SymmetricCipherMode mode,
+                     SymmetricCipherPadding padding,
+                     const std::vector<uint8_t> &secretKey,
+                     const std::vector<uint8_t> &iv,
+                     Operation operation)
+{
+    _impl = std::make_unique<AESCipher::Impl>(mode, padding, secretKey, iv, operation);
+}
+
+void AESCipher::update(const std::vector<uint8_t> &message) { _impl->update(message); }
+
+std::vector<uint8_t> AESCipher::read(size_t length) { return _impl->read(length); }
+
+std::vector<uint8_t> AESCipher::readAll() { return _impl->readAll(); }
+
+std::vector<uint8_t> AESCipher::finish()
+{
+    return _impl->finish();
+}
+
+std::vector<uint8_t> AESCipher::getIV() { return _impl->getIV(); }
+
+
+AuthenticatedAESCipher::AuthenticatedAESCipher(SymmetricCipherMode mode,
+                                               SymmetricCipherPadding padding,
+                                               const std::vector<uint8_t> &secretKey,
+                                               const std::vector<uint8_t> &iv,
+                                               const std::vector<uint8_t> &authTag,
+                                               AESCipher::Operation operation)
+        : AESCipher(mode, padding, secretKey, iv, operation)
+{
+    _impl->setAuthTag(authTag);
+}
+
+std::vector<uint8_t> AuthenticatedAESCipher::getAuthTag()
+{
+    return _impl->getAuthTag();
+}
+
+void AuthenticatedAESCipher::setAuthTag(const std::vector<uint8_t> &tag)
+{
+    _impl->setAuthTag(tag);
+}
+
+
+AESCipherBuilder::AESCipherBuilder(SymmetricCipherMode mode,
+                                   std::vector<uint8_t> secretKey)
+        : _mode{mode}, _secretKey{std::move(secretKey)}
+{
+}
+
+AESCipherBuilder &AESCipherBuilder::setIV(std::vector<uint8_t> const &iv)
```

Comment:
Rather than copying, should we outright accept copies of the iv and `std::move()` them, first into the `AESCipherBuilder`, and potentially later directly into the `AESCipher` instance?